### DISCUSSION TITLE
Need some help to get I2C work with a DS3231 board

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#134] [General]
I am trying to get a Pi Pico to talk to a DS3231 RTC breakout board. 
Using I2C0 GPIO 4 & 5 and the i2c scanner of the test-sources of Zeptoforth confirms
that the DS3231 is found on $57 and $68. Since I found more sources only, which uses $68 
I used it too.

The source I have so far (its a bunch of words to test and experiment right now)

```
\ DS3231 i2c driver

defined? ds3231 [if] ds3231 [then]
marker ds3231


begin-module ds3231drv

pin import
i2c import

16 buffer: i2cbuf

0 constant i2cdev
\ 400000 constant i2cclock
100000 constant i2cclock
4 constant sda0pin
5 constant scl0pin

: i2cbuf-clr
  0 i2cbuf  0 + !
  0 i2cbuf  4 + !
  0 i2cbuf  8 + !
  0 i2cbuf 12 + !
;


: i2c-setup
  i2cbuf-clr
  i2cdev clear-i2c
  i2cdev sda0pin i2c-pin
  i2cdev scl0pin i2c-pin
  i2cclock i2cdev i2c-clock!
  i2cdev master-i2c
  i2cdev 7-bit-i2c-addr
  $68 i2cdev i2c-slave-addr!
  i2cdev enable-i2c
;

: ds-sendres { n }
  i2cbuf n i2cdev >i2c-restart
;

: ds-send { n }
  i2cbuf n i2cdev >i2c
;

: dsdump
  i2cbuf 20 xdump
;

: i2cbuf-set ( val0...valn n - )
  0 do i2cbuf i + c! loop
;

: ds-rec ( -- )
  i2cbuf-clr
  dsdump
;

: ds-rw0
  \ ." A"
  \ i2cbuf-clr
  \ $68 i2cbuf !
  \ i2cbuf 1 i2cdev >i2c
  ." B"
  i2cbuf-clr
  $0 i2cbuf !
  i2cbuf 1 i2cdev >i2c-restart-stop
  .
  ." C"
  i2cbuf-clr
  i2cbuf 4 i2cdev i2c-stop>
  dsdump
;

end-module

ds3231drv import
i2c-setup
```
I do not need any high speed handling here, since the "measured value" only changes once per second.
The last thing I tried, was the word ds-rw0, but as soon, it tries to send anything, the Pico locks up hard
and needs a reset via button press.

What seems to work is sending a byte to the DS3231, as "1" as the number of bytes sent,  is given back.

I did quite a number of variations of >i2c and i2c>, which I thought are in line with the DS3231 datasheet,
but all result in the same: Byte 1 send: OK, trying to send a 2nd byte: hangs, receiving any byte: hangs,

...and I think, I am also stuck...

Any help is very appreciated.

**Comment by tabemann** [comment]
On Sat, Aug 17, 2024 at 4:02 AM mcctuxic ***@***.***> wrote:

> I am trying to get a Pi Pico to talk to a DS3231 RTC breakout board.
> Using I2C0 GPIO 4 & 5 and the i2c scanner of the test-sources of
> Zeptoforth confirms
> that the DS3231 is found on $57 and $68. Since I found more sources only,
> which uses $68
> I used it too.
>
> The source I have so far (its a bunch of words to test and experiment
> right now)
>
> \ DS3231 i2c driver
>
> defined? ds3231 [if] ds3231 [then]
> marker ds3231
>
>
> begin-module ds3231drv
>
> pin import
> i2c import
>
> 16 buffer: i2cbuf
>
> 0 constant i2cdev
> \ 400000 constant i2cclock
> 100000 constant i2cclock
> 4 constant sda0pin
> 5 constant scl0pin
>
> : i2cbuf-clr
>   0 i2cbuf  0 + !
>   0 i2cbuf  4 + !
>   0 i2cbuf  8 + !
>   0 i2cbuf 12 + !
> ;
>
>
> : i2c-setup
>   i2cbuf-clr
>   i2cdev clear-i2c
>   i2cdev sda0pin i2c-pin
>   i2cdev scl0pin i2c-pin
>   i2cclock i2cdev i2c-clock!
>   i2cdev master-i2c
>   i2cdev 7-bit-i2c-addr
>   $68 i2cdev i2c-slave-addr!
>   i2cdev enable-i2c
> ;
>
> : ds-sendres { n }
>   i2cbuf n i2cdev >i2c-restart
> ;
>
> : ds-send { n }
>   i2cbuf n i2cdev >i2c
> ;
>
> : dsdump
>   i2cbuf 20 xdump
> ;
>
> : i2cbuf-set ( val0...valn n - )
>   0 do i2cbuf i + c! loop
> ;
>
> : ds-rec ( -- )
>   i2cbuf-clr
>   dsdump
> ;
>
> : ds-rw0
>   \ ." A"
>   \ i2cbuf-clr
>   \ $68 i2cbuf !
>   \ i2cbuf 1 i2cdev >i2c
>   ." B"
>   i2cbuf-clr
>   $0 i2cbuf !
>   i2cbuf 1 i2cdev >i2c-restart-stop
>   .
>   ." C"
>   i2cbuf-clr
>   i2cbuf 4 i2cdev i2c-stop>
>   dsdump
> ;
>
> end-module
>
> ds3231drv import
> i2c-setup
>
> I do not need any high speed handling here, since the "measured value"
> only changes once per second.
> The last thing I tried, was the word ds-rw0, but as soon, it tries to send
> anything, the Pico locks up hard
> and needs a reset via button press.
>
> What seems to work is sending a byte to the DS3231, as "1" as the number
> of bytes sent, is given back.
>
> I did quite a number of variations of >i2c and i2c>, which I thought are
> in line with the DS3231 datasheet,
> but all result in the same: Byte 1 send: OK, trying to send a 2nd byte:
> hangs, receiving any byte: hangs,
>
> ...and I think, I am also stuck...
>
> Any help is very appreciated.
>

One thing that needs changing is $68 I2C-SLAVE-ADDR! - this should be $68
I2C-TARGET-ADDR! because you are a master and not a slave. Other than that
there is nothing obviously wrong with your code. (Note that you did not
include certain words you are using, so I do not know if you were using
them properly.)

I made a simple test where I have a board (in this case a Pico W, as I had
one that was already wired up to talk between GPIO's 12 and 13 as I2C0 as a
master and GPIO's 14 and 15 as I2C1 as a slave with pull-up resistors) talk
to itself without a problem. When running this test load respond.fs first
and master.fs second. It works just fine, with the contents of the message
sent from the slave to the master being dumped out properly.

Travis

**Comment by tabemann** [comment]
On Sat, Aug 17, 2024 at 7:16 PM Travis Bemann ***@***.***> wrote:

> On Sat, Aug 17, 2024 at 4:02 AM mcctuxic ***@***.***> wrote:
>
>> I am trying to get a Pi Pico to talk to a DS3231 RTC breakout board.
>> Using I2C0 GPIO 4 & 5 and the i2c scanner of the test-sources of
>> Zeptoforth confirms
>> that the DS3231 is found on $57 and $68. Since I found more sources only,
>> which uses $68
>> I used it too.
>>
>> The source I have so far (its a bunch of words to test and experiment
>> right now)
>>
>> \ DS3231 i2c driver
>>
>> defined? ds3231 [if] ds3231 [then]
>> marker ds3231
>>
>>
>> begin-module ds3231drv
>>
>> pin import
>> i2c import
>>
>> 16 buffer: i2cbuf
>>
>> 0 constant i2cdev
>> \ 400000 constant i2cclock
>> 100000 constant i2cclock
>> 4 constant sda0pin
>> 5 constant scl0pin
>>
>> : i2cbuf-clr
>>   0 i2cbuf  0 + !
>>   0 i2cbuf  4 + !
>>   0 i2cbuf  8 + !
>>   0 i2cbuf 12 + !
>> ;
>>
>>
>> : i2c-setup
>>   i2cbuf-clr
>>   i2cdev clear-i2c
>>   i2cdev sda0pin i2c-pin
>>   i2cdev scl0pin i2c-pin
>>   i2cclock i2cdev i2c-clock!
>>   i2cdev master-i2c
>>   i2cdev 7-bit-i2c-addr
>>   $68 i2cdev i2c-slave-addr!
>>   i2cdev enable-i2c
>> ;
>>
>> : ds-sendres { n }
>>   i2cbuf n i2cdev >i2c-restart
>> ;
>>
>> : ds-send { n }
>>   i2cbuf n i2cdev >i2c
>> ;
>>
>> : dsdump
>>   i2cbuf 20 xdump
>> ;
>>
>> : i2cbuf-set ( val0...valn n - )
>>   0 do i2cbuf i + c! loop
>> ;
>>
>> : ds-rec ( -- )
>>   i2cbuf-clr
>>   dsdump
>> ;
>>
>> : ds-rw0
>>   \ ." A"
>>   \ i2cbuf-clr
>>   \ $68 i2cbuf !
>>   \ i2cbuf 1 i2cdev >i2c
>>   ." B"
>>   i2cbuf-clr
>>   $0 i2cbuf !
>>   i2cbuf 1 i2cdev >i2c-restart-stop
>>   .
>>   ." C"
>>   i2cbuf-clr
>>   i2cbuf 4 i2cdev i2c-stop>
>>   dsdump
>> ;
>>
>> end-module
>>
>> ds3231drv import
>> i2c-setup
>>
>> I do not need any high speed handling here, since the "measured value"
>> only changes once per second.
>> The last thing I tried, was the word ds-rw0, but as soon, it tries to
>> send anything, the Pico locks up hard
>> and needs a reset via button press.
>>
>> What seems to work is sending a byte to the DS3231, as "1" as the number
>> of bytes sent, is given back.
>>
>> I did quite a number of variations of >i2c and i2c>, which I thought are
>> in line with the DS3231 datasheet,
>> but all result in the same: Byte 1 send: OK, trying to send a 2nd byte:
>> hangs, receiving any byte: hangs,
>>
>> ...and I think, I am also stuck...
>>
>> Any help is very appreciated.
>>
>
> One thing that needs changing is $68 I2C-SLAVE-ADDR! - this should be $68
> I2C-TARGET-ADDR! because you are a master and not a slave. Other than that
> there is nothing obviously wrong with your code. (Note that you did not
> include certain words you are using, so I do not know if you were using
> them properly.)
>
> I made a simple test where I have a board (in this case a Pico W, as I had
> one that was already wired up to talk between GPIO's 12 and 13 as I2C0 as a
> master and GPIO's 14 and 15 as I2C1 as a slave with pull-up resistors) talk
> to itself without a problem. When running this test load respond.fs first
> and master.fs second. It works just fine, with the contents of the message
> sent from the slave to the master being dumped out properly.
>

Note that my example does not need a Pico W to run. It should run on a
plain old Pico just fine with the proper pull-up resistors.

Travis

**Comment by mcctuxic** [comment]
On 08/17 05:22, tabemann wrote:
> On Sat, Aug 17, 2024 at 7:16 PM Travis Bemann ***@***.***> wrote:
> 
> > On Sat, Aug 17, 2024 at 4:02 AM mcctuxic ***@***.***> wrote:
> >
> >> I am trying to get a Pi Pico to talk to a DS3231 RTC breakout board.
> >> Using I2C0 GPIO 4 & 5 and the i2c scanner of the test-sources of
> >> Zeptoforth confirms
> >> that the DS3231 is found on $57 and $68. Since I found more sources only,
> >> which uses $68
> >> I used it too.
> >>
> >> The source I have so far (its a bunch of words to test and experiment
> >> right now)
> >>
> >> \ DS3231 i2c driver
> >>
> >> defined? ds3231 [if] ds3231 [then]
> >> marker ds3231
> >>
> >>
> >> begin-module ds3231drv
> >>
> >> pin import
> >> i2c import
> >>
> >> 16 buffer: i2cbuf
> >>
> >> 0 constant i2cdev
> >> \ 400000 constant i2cclock
> >> 100000 constant i2cclock
> >> 4 constant sda0pin
> >> 5 constant scl0pin
> >>
> >> : i2cbuf-clr
> >>   0 i2cbuf  0 + !
> >>   0 i2cbuf  4 + !
> >>   0 i2cbuf  8 + !
> >>   0 i2cbuf 12 + !
> >> ;
> >>
> >>
> >> : i2c-setup
> >>   i2cbuf-clr
> >>   i2cdev clear-i2c
> >>   i2cdev sda0pin i2c-pin
> >>   i2cdev scl0pin i2c-pin
> >>   i2cclock i2cdev i2c-clock!
> >>   i2cdev master-i2c
> >>   i2cdev 7-bit-i2c-addr
> >>   $68 i2cdev i2c-slave-addr!
> >>   i2cdev enable-i2c
> >> ;
> >>
> >> : ds-sendres { n }
> >>   i2cbuf n i2cdev >i2c-restart
> >> ;
> >>
> >> : ds-send { n }
> >>   i2cbuf n i2cdev >i2c
> >> ;
> >>
> >> : dsdump
> >>   i2cbuf 20 xdump
> >> ;
> >>
> >> : i2cbuf-set ( val0...valn n - )
> >>   0 do i2cbuf i + c! loop
> >> ;
> >>
> >> : ds-rec ( -- )
> >>   i2cbuf-clr
> >>   dsdump
> >> ;
> >>
> >> : ds-rw0
> >>   \ ." A"
> >>   \ i2cbuf-clr
> >>   \ $68 i2cbuf !
> >>   \ i2cbuf 1 i2cdev >i2c
> >>   ." B"
> >>   i2cbuf-clr
> >>   $0 i2cbuf !
> >>   i2cbuf 1 i2cdev >i2c-restart-stop
> >>   .
> >>   ." C"
> >>   i2cbuf-clr
> >>   i2cbuf 4 i2cdev i2c-stop>
> >>   dsdump
> >> ;
> >>
> >> end-module
> >>
> >> ds3231drv import
> >> i2c-setup
> >>
> >> I do not need any high speed handling here, since the "measured value"
> >> only changes once per second.
> >> The last thing I tried, was the word ds-rw0, but as soon, it tries to
> >> send anything, the Pico locks up hard
> >> and needs a reset via button press.
> >>
> >> What seems to work is sending a byte to the DS3231, as "1" as the number
> >> of bytes sent, is given back.
> >>
> >> I did quite a number of variations of >i2c and i2c>, which I thought are
> >> in line with the DS3231 datasheet,
> >> but all result in the same: Byte 1 send: OK, trying to send a 2nd byte:
> >> hangs, receiving any byte: hangs,
> >>
> >> ...and I think, I am also stuck...
> >>
> >> Any help is very appreciated.
> >>
> >
> > One thing that needs changing is $68 I2C-SLAVE-ADDR! - this should be $68
> > I2C-TARGET-ADDR! because you are a master and not a slave. Other than that
> > there is nothing obviously wrong with your code. (Note that you did not
> > include certain words you are using, so I do not know if you were using
> > them properly.)
> >
> > I made a simple test where I have a board (in this case a Pico W, as I had
> > one that was already wired up to talk between GPIO's 12 and 13 as I2C0 as a
> > master and GPIO's 14 and 15 as I2C1 as a slave with pull-up resistors) talk
> > to itself without a problem. When running this test load respond.fs first
> > and master.fs second. It works just fine, with the contents of the message
> > sent from the slave to the master being dumped out properly.
> >
> 
> Note that my example does not need a Pico W to run. It should run on a
> plain old Pico just fine with the proper pull-up resistors.
> 
> Travis
> 
> > Message ID: ***@***.***>
> >>
> >

Hi Travis,

Thank you for your help! :)

:) I am using an Pico W also...and only forgot to add the "W" in my post :)

Everything is fine now as it seems: I can read values from the DS3231
RTC and they are changing over time. Looks good!

The reason, why it does not work previously seems to be a fried 
I2C-port 0 at GP4/GP5. I changed to port 1 and GP6/GP7 then.

I installed Micropython temporarily and could confirm, that a read from
the fried port "hangs" the system. I changed to the other port than, and
everything works fine.

**Comment by tabemann** [comment]
On Sun, Aug 18, 2024, 03:34 mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> Thank you for your help! :)
>
> :) I am using an Pico W also...and only forgot to add the "W" in my post :)
>
> Everything is fine now as it seems: I can read values from the DS3231
> RTC and they are changing over time. Looks good!
>
> The reason, why it does not work previously seems to be a fried
> I2C-port 0 at GP4/GP5. I changed to port 1 and GP6/GP7 then.
>
> I installed Micropython temporarily and could confirm, that a read from
> the fried port "hangs" the system. I changed to the other port than, and
> everything works fine.
>

Good to hear that you got it working, and sorry to hear about your
partially fried Pico W!

Travis

>

**Comment by mcctuxic** [comment]
No need to be sorry, Travis! I am happy, that it is working. One step closer to the nixie clock ;)

**Comment by tabemann** [comment]
On Sun, Aug 18, 2024, 12:27 mcctuxic ***@***.***> wrote:

> No need to be sorry, Travis! I am happy, that it is working. One step
> closer to the nixie clock ;)
>
One thing I would suggest is using another Pico W because if I2C0 is fried
who knows what else is also bad.

Travis

**Comment by mcctuxic** [comment]
On 08/18 10:36, tabemann wrote:
> On Sun, Aug 18, 2024, 12:27 mcctuxic ***@***.***> wrote:
> 
> > No need to be sorry, Travis! I am happy, that it is working. One step
> > closer to the nixie clock ;)
> >
> One thing I would suggest is using another Pico W because if I2C0 is fried
> who knows what else is also bad.
> 
> Travis

GOOD IDEA! Will do so - thank you! :)
